File: src/visual-graph/builder/is-branch-scope.test.ts
import { describe, expect, test } from "vitest";

import type { BlockContext } from "../../ir/scope/block-context.js";
import { asScopeId } from "../../ir/serialized/scope-id.js";
import { AST_TYPE } from "../../parser/ast-type.js";
import { asFilledString } from "../../util/filled-string.js";
import { isBranchScope } from "./is-branch-scope.js";
import { baseBlockContext } from "./testing/make-block-context.js";
import { baseScope } from "./testing/make-scope.js";

describe("isBranchScope", () => {
  test.each<{ name: string; ctx: BlockContext | null; expected: boolean }>([
    {
      name: asFilledString("if consequent block scope -> true"),
      ctx: {
        ...baseBlockContext(),
        parentType: AST_TYPE.IfStatement,
        key: asFilledString("consequent"),
        parentSpanOffset: 0,
      },
      expected: true,
    },
    {
      name: asFilledString("if alternate block scope -> true"),
      ctx: {
        ...baseBlockContext(),
        parentType: AST_TYPE.IfStatement,
        key: asFilledString("alternate"),
        parentSpanOffset: 0,
      },
      expected: true,
    },
    {
      name: asFilledString("switch case scope -> true"),
      ctx: {
        ...baseBlockContext(),
        parentType: AST_TYPE.SwitchStatement,
        key: asFilledString("cases"),
        parentSpanOffset: 0,
      },
      expected: true,
    },
    {
      name: asFilledString(
        "try block scope -> true (try and catch are sibling branches)",
      ),
      ctx: {
        ...baseBlockContext(),
        parentType: AST_TYPE.TryStatement,
        key: asFilledString("block"),
        parentSpanOffset: 0,
      },
      expected: true,
    },
    {
      name: asFilledString(
        "try handler scope -> true (try and catch are sibling branches)",
      ),
      ctx: {
        ...baseBlockContext(),
        parentType: AST_TYPE.TryStatement,
        key: asFilledString("handler"),
        parentSpanOffset: 0,
      },
      expected: true,
    },
    {
      name: asFilledString(
        "try finalizer scope -> false (finally is post-merge, not branch)",
      ),
      ctx: {
        ...baseBlockContext(),
        parentType: AST_TYPE.TryStatement,
        key: asFilledString("finalizer"),
        parentSpanOffset: 0,
      },
      expected: false,
    },
    {
      name: asFilledString("no blockContext -> false"),
      ctx: null,
      expected: false,
    },
  ])("$name", ({ ctx, expected }) => {
    const scope = { ...baseScope(), id: asScopeId("s"), blockContext: ctx };
    const map = new Map([[scope.id, scope]]);
    expect(isBranchScope("s", map)).toEqual(expected);
  });

  test("scope id missing from the map -> false", () => {
    expect(isBranchScope("missing", new Map())).toEqual(false);
  });
});

File: crates/unsnarl-visual-graph/src/builder/is_branch_scope_test.rs
//! Sibling tests for [`is_branch_scope`]. Cases mirror
//! `ts/src/visual-graph/builder/is-branch-scope.test.ts`.

use std::collections::HashMap;

use unsnarl_ir::scope::block_context::BlockContext;
use unsnarl_ir::serialized::SerializedScope;
use unsnarl_oxc_parity::AstType;

use super::is_branch_scope;
use crate::builder::testing::{base_serialized_scope, other_block_context};

fn scope_with_context(ctx: Option<BlockContext>) -> SerializedScope {
    let mut scope = base_serialized_scope("s");
    scope.block_context = ctx;
    scope
}

fn check(ctx: Option<BlockContext>) -> bool {
    let scope = scope_with_context(ctx);
    let mut map: HashMap<&str, &SerializedScope> = HashMap::new();
    map.insert(scope.id.value(), &scope);
    is_branch_scope("s", &map)
}

#[test]
fn if_consequent_block_scope_is_branch() {
    let ctx = other_block_context(AstType::IfStatement, "consequent", 0, None);
    assert!(check(Some(ctx)));
}

#[test]
fn if_alternate_block_scope_is_branch() {
    let ctx = other_block_context(AstType::IfStatement, "alternate", 0, None);
    assert!(check(Some(ctx)));
}

#[test]
fn switch_case_scope_is_branch() {
    // is_branch_scope routes through `branch_container_key`, which
    // for the `switch:` case key requires the `CaseClause` variant.
    // The TS test uses a generic block context, but the Rust impl
    // separates the variants — exercise the `CaseClause` variant
    // here so the test mirrors the underlying merge-key contract.
    use unsnarl_ir::primitive::SourceOffset;
    use unsnarl_ir::scope::block_context::CaseClauseBlockContext;
    let ctx = BlockContext::CaseClause(CaseClauseBlockContext::new(
        AstType::SwitchStatement,
        "cases".to_string(),
        SourceOffset(0),
        None,
    ));
    assert!(check(Some(ctx)));
}

#[test]
fn try_block_scope_is_branch() {
    // `try` and `catch` are sibling branches.
    let ctx = other_block_context(AstType::TryStatement, "block", 0, None);
    assert!(check(Some(ctx)));
}

#[test]
fn try_handler_scope_is_branch() {
    // `try` and `catch` are sibling branches.
    let ctx = other_block_context(AstType::TryStatement, "handler", 0, None);
    assert!(check(Some(ctx)));
}

#[test]
fn try_finalizer_scope_is_not_branch() {
    // `finally` is post-merge, not a branch.
    let ctx = other_block_context(AstType::TryStatement, "finalizer", 0, None);
    assert!(!check(Some(ctx)));
}

#[test]
fn scope_without_block_context_is_not_branch() {
    assert!(!check(None));
}

#[test]
fn scope_id_missing_from_map_is_not_branch() {
    let map: HashMap<&str, &SerializedScope> = HashMap::new();
    assert!(!is_branch_scope("missing", &map));
}
